### ISSUE TITLE
ggml-hexagon on Hexagon v73 (QCM6690): NPU output is byte-identical regardless of the prompt — silent corruption, no crash, `graph splits = 3`

### ISSUE THREAD

**Issue by sGeeK44** [qualcomm/GenieX#1429]
## Describe the bug

On a Zebra TC501 (Qualcomm QCM6690, Hexagon **v73**), the `llama_cpp` plugin with `--device npu` produces output that **does not depend on the input prompt at all**. Two completely different prompts yield byte-identical generations. The same binary and model on `--device cpu` answers both correctly.

There is no crash, no error and no fallback warning: the model loads, all layers are offloaded to `HTP0`, `graph splits = 3`, throughput is normal. The results are simply invalid.

This is distinct from the existing crash reports (#1170 on V68, #1390 on X Elite) — nothing fails here, the output is silently wrong.

## To Reproduce

```bash
# geniex-bench v0.6.1, android-arm64
adb push bin lib /data/local/tmp/gx061/
adb shell cp /data/local/tmp/gx061/lib/llama_cpp/libggml-htp-v73.so /data/local/tmp/gx061/lib/

printf 'What is the capital of France?'  > neutral.txt   # 30 bytes
printf 'Quelle est la capitale de la France ?' > other.txt # different prompt, different length
adb push neutral.txt other.txt gemma-3-1b-Q4_0.gguf /data/local/tmp/

run() {  # $1 = device, $2 = prompt file
  adb shell "cd /data/local/tmp && \
    LD_LIBRARY_PATH=/data/local/tmp/gx061/lib:/data/local/tmp/gx061/lib/llama_cpp \
    ADSP_LIBRARY_PATH=/data/local/tmp/gx061/lib \
    /data/local/tmp/gx061/bin/geniex-bench --plugin llama_cpp --device $1 \
      -m gemma-3-1b-Q4_0.gguf --prompt-file $2 --accuracy -n 30" | grep '^\[gen '
}

run npu neutral.txt > npu_a.txt ; run npu other.txt > npu_b.txt ; diff npu_a.txt npu_b.txt
run cpu neutral.txt > cpu_a.txt ; run cpu other.txt > cpu_b.txt ; diff cpu_a.txt cpu_b.txt
```

## Expected behavior

`--device npu` answers the prompt, and different prompts give different outputs — as `--device cpu` does.

## Actual behavior

**`--device npu`** — the two prompts produce **identical bytes**:

```
$ diff npu_a.txt npu_b.txt          # no output
$ sha256sum npu_a.txt npu_b.txt
2352d6398a4142043c79f4ac99e3d3c5b7887d5c67a79e907e0f69f64fddecc1  npu_a.txt
2352d6398a4142043c79f4ac99e3d3c5b7887d5c67a79e907e0f69f64fddecc1  npu_b.txt
```

and the content itself is random tokens across unrelated scripts:

```
[gen ] 𒐩ც千萬ङ् đối dibuka违法 afford eigenvalue альбо LHຍ -லிக்கDimsbewer menambahकडील дослідজিStudioProjectsühm姓ствииতই dessasまず ஆகும்ให้럽 αποκ导师 vertebrateAwayмя razón mujer kickedőség4 हल्की equalizer प्लानдамbs7тил टीकेüte here
[ok  ] cell plugin=llama_cpp device=npu(id=HTP0) ngl=-1 ttft=136.0ms prefill=119.1tps decode=14.8tps gen=50 tok
```

**`--device cpu`** — same two prompts, correct and distinct answers:

```
$ cat cpu_a.txt
[gen ] The capital of France is Paris.
$ cat cpu_b.txt
[gen ] La capitale de la France est Paris.
$ sha256sum cpu_a.txt cpu_b.txt
3765449a6db65bfc0f7c48c8e162b57ee23515c5341bf94c943c6bd67f5a6e25  cpu_a.txt
e5c3ae5c5e83465cb8af55d43bf5f0ba3e0d7d7ccf5d460ec41b99cf7c8e24a5  cpu_b.txt
```

The CPU run is the control: it proves the harness does feed the prompt through, and that only the NPU path discards it.

## Scope — what we ruled out

Same device, same binary; only `--device` and the model change.

| Model | Quant | `cpu` | `npu` | `hybrid` |
|---|---|---|---|---|
| `Gemma-3-1B-It` (gemma3) | Q4_0 | correct | random multi-script tokens | — |
| `Qwen3-1.7B` (qwen3) | Q8_0 | correct | ` subsidi9855155131671 / 3213333713450572217463349` | `.+.+.+.+.+.+.+.+.+.+` |

So it is **not** specific to:

- **the architecture** — `gemma3` and `qwen3` fail identically;
- **the quantization** — `Q4_0` and `Q8_0` both fail;
- **the prompt** — a 30-byte prompt fails, and the output is prompt-independent, so no context/VTCM threshold is involved;
- **the release** — reproduced on **v0.5.0** and **v0.6.1**;
- **the compute unit alias** — `hybrid` is also wrong (and much slower: `ttft=13634.8ms`, `prefill=1.1tps`).

`Qwen3-1.7B` in `Q4_0` fails the same way when driven through the Android SDK (`LlmWrapper`) on the same device.

Failure modes differ slightly between models: Gemma emits token-level noise, while Qwen3 sometimes emits fluent, grammatical text unrelated to the prompt. Combined with the prompt-independence, that looks like decode running on an empty/stale context rather than pure logit noise — which would point at the prefill path.

## Environment

| | |
|---|---|
| Device | Zebra TC501 (`ro.product.model=TC501`) |
| SoC | `ro.soc.model=QCM6690`, `ro.board.platform=volcano` |
| Hexagon | `ggml-hex: Hexagon Arch version v73` |
| HTP hwinfo | `threads 2, hvx 2, hmx 1, vtcm 2 MB` |
| OS | Android 15 — `Zebra/TC501W/TCX01W:15/15-16-22.00-VG-U00-PRD-ERS-04/371:user/release-keys` |
| GenieX | `geniex-bench` v0.6.1 android-arm64 (also reproduced on v0.5.0) |
| Backend | `llama_cpp` plugin, `ggml-hexagon` |

Models:

- `gemma-3-1b-Q4_0.gguf` — `general.repo_url = https://huggingface.co/unsloth`, `general.name = Gemma-3-1B-It`, 721 918 496 bytes, sha256 `27ee88e03be02e9ba73def9a819d570d8ad73716e50769e87f374ae394b0276e`
- `Qwen3-1.7B-Q8_0.gguf` — 1 834 426 944 bytes, sha256 `0becaa825564295d82e9af4d008bca5f8b7f5f73bf1c6a0b58f7c53ef26b47fd`

## Logs (`GENIEX_LOG=debug`, `--device npu`)

```
ggml-hex: Loading driver libcdsprpc.so
ggml-hex: forcing ndev to 1 for SoCs archs lower than v75.
ggml-hex: FASTRPC_GET_DOMAINS query failed (0x6c), using static CDSP domains
ggml-hex: Hexagon backend (experimental) : allocating new registry : ndev 1
ggml-hex: Hexagon Arch version v73
ggml-hex: HTP0 allocating new session
ggml-hex: HTP0 hwinfo: threads 2, hvx 2, hmx 1, vtcm 2 MB
ggml-hex: HTP0 new session : session-id 0 domain-id 3 uri file:///libggml-htp-v73.so?...&_dom=cdsp&_session=0
ggml-hex: HTP0 op batching: n-bufs 16 n-tensors 8192 n-ops 1024 vmem 3145728000
load_tensors: offloaded 27/27 layers to GPU
sched_reserve: graph nodes  = 1047
sched_reserve: graph splits = 3
llama_context: n_ctx = 4096, flash_attn_type = 1
```

Nothing in the log signals a problem: `graph splits = 3`, all 27 layers on `HTP0`. The scheduler believes the graph runs on the NPU as intended.

## Additional context

Two v73-specific paths show up during init and may be worth a look:

- `forcing ndev to 1 for SoCs archs lower than v75` — v73 takes a dedicated path here.
- `FASTRPC_GET_DOMAINS query failed (0x6c), using static CDSP domains` — domain discovery falls back to static values on this device.

We have this hardware available and are happy to run further diagnostics (e.g. `--logits` dumps for a CPU/NPU comparison, or per-op bisection) if that helps narrow it down.

**Comment by john-rocky** [comment]
Could you run the attached small capture on the TC501 using your existing v0.6.1 package and the two GGUF files? It collects the formatted prompt, actual backend input token IDs, unfiltered CPU/NPU output, binary/model hashes, and the first four full-vocabulary logit rows. It compares numerical values only while the complete input token prefixes match.

From the extracted directory, replace `TC501_ADB_SERIAL` and adjust only the existing device paths if needed (Python 3.11+ and adb on the host):

```sh
python3 run.py --serial TC501_ADB_SERIAL \
  --package /data/local/tmp/gx061 \
  --trace ./libgx1429.so \
  --model /data/local/tmp/gemma-3-1b-Q4_0.gguf \
  --model /data/local/tmp/Qwen3-1.7B-Q8_0.gguf \
  --stock-control \
  --out ./tc501-1429
```

Each run starts fresh, uses 30 tokens maximum and greedy sampling (`temperature=-1`; this plugin treats `0` as an unset/default value). The two default prompts ask for France's capital and 2+2. Source and rebuild instructions are included; no model download or SDK rebuild is needed.

Please return the result directory, including `manifest.json`, stdout/stderr, `trace.jsonl`, and the `.f32` rows. If capture fails, please retain the partial results. The stock controls help identify whether tracing changes the symptom.

I verified collection on SM8850/v81 with a different Qwen3-0.6B Q4_K_M model and substantial CPU fallback. That does not validate or refute the TC501/v73 report; I have not reproduced it on your hardware.

Attachment: [geniex-1429-diagnostic.zip](https://github.com/user-attachments/files/31865100/geniex-1429-diagnostic.zip) (33 KB).

SHA-256: `e2527f8487b36e1c09e604a0dfae334e210a0a3c63c6b4b9ab7310cc4630d0df`

**Comment by mansiverma897993** [comment]
@mengshengwu I'd like to try to implement a fix for this v73 issue, but I don't have a QCM6690 / Zebra TC501 device here, so I can't reproduce it myself. Would one of the you (or anyone with v73 hardware) be able to reproduce/verify my PR, or run the diagnostic capture on this device? I can prepare the patch and CI-safe unit checks on my side first. Thanks! @vinovo then I will start working here !!

**Comment by RichardDong1225** [comment]
Hi,
It's great to receive your feedback. We've logged your request and will provide updates under this issue as progress is made.
BR
